docs/devel/consumers-claims.md

```diff
@@ -0,0 +1,309 @@
+        |              +---------------+
+        +---watch----> | IngressClaim  | <--+
+                       +---------------+    |
+                                 ^          |
+                                 |          +-------- user-created
+                                 |          |
+                           +----------+     |
+                           | Ingress  | <---+
+                           +----------+
+```
+
+### IngressClaim
+
+An `IngressClaim` does not specify which specific Ingress it wants.  Instead it
+specifies characteristics of the ingress it needs and those characteristics are
+satisfied as best is possible.  Characteristics might include the availability
+properties required or the specific IP address to use.
+
+Like `PersistentVolumeClaim`, users will be able to request coarse "classes" of
+ingress (e.g. Gold, Silver, Bronze).  How those classes map to real storage
+technologies is a policy set by the cluster administrators.
+
+### LoadBalancerManagers
+
+The act of binding a `IngressClaim` to an `Ingress` is the responsibility of a
+process in the cluster.  A cluster may have any number of
+`LoadBalancerManager`s running, each of which handles some set of claim
+classes.  When an `IngressClaim` is created that asks for "Gold" ingress, the
+`LoadBalancerManager` which understands the "Gold" class will make the binding.
+
+`LoadBalancerManager`s can be simple `Pod`s running in the cluster, pulling
+cofiguration from hardcoded values, flags, config objects, or whatever
+mechanism makes sense for them.  We expect config objects to be a perfect fit
+for this use case.
+
+This very loose coupling means that an admin can define classes however they
+like.  For example, "Gold" might mean GCE Cloud Load Balancer, "Silver"
+might mean HAProxy, and "Bronze" might mean a port on a single VM.
+Adding a totally bespoke `LoadBalancerManager` is trivial.
+
+Note that classes are just identifiers - there are no special names or
+meanings.  One of the `LoadBalancerManager`s must be the "default" for claims that do
+not specify a class.
+
+### Config
+
+We expect that `LoadBalancerManager`s will need some amount of configuration, and it
+will be convenient if they have similar parameters, but this should be by
+convention, not by diktat.  Example parameters:
+   * Max instances - limit of number of balancers allowed
+   * Min free instances - how many ready-to-use balancers should be maintained
+   * Max free instances - limit of number of ready-to-use balancers
+
+### Scrubbing
+
+While it's a bit more of a stretch, even load-balancers have a conceptual
+scrubbing phase.  For example, if a particular IP address was used with a DNS name, the
+system might choose not to reuse that IP until the TTL of the DNS name has
+expired.
+
+## Future
+
+We hope that this pattern and the commonalities continue to hold over time.  As
+we integrate new cluster resources, we hope this model will also extend to
+them.
+
```

Comment:
And the need for status to be indicated to clearly communicate claim
status.  Also that claims being one way has advantages for many models (pod
-> node, PVC -> pv) are all at most once and last until deletion which
simplifies user expectation.

On Sep 17, 2015, at 6:47 PM, Prashanth B <notifications@github.com> wrote:

In docs/devel/consumers-claims.md
<https://github.com/kubernetes/kubernetes/pull/14095#discussion_r39808210>:

> +   * Min free instances - how many ready-to-use balancers should be maintained
> +   * Max free instances - limit of number of ready-to-use balancers
> +
> +### Scrubbing
> +
> +While it's a bit more of a stretch, even load-balancers have a conceptual
> +scrubbing phase.  For example, if a particular IP address was used with a DNS name, the
> +system might choose not to reuse that IP until the TTL of the DNS name has
> +expired.
> +
> +## Future
> +
> +We hope that this pattern and the commonalities continue to hold over time.  As
> +we integrate new cluster resources, we hope this model will also extend to
> +them.
> +

Mention limitations of lose coupling somewhere? (nebulous failures,
undefined claim etc)

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/14095/files#r39808210>.